depends/toolchain.cmake.in

```diff
@@ -4,8 +4,10 @@
 
 # This file is expected to be highly volatile and may still change substantially.
 
-set(CMAKE_SYSTEM_NAME @host_system@)
-set(CMAKE_SYSTEM_PROCESSOR @host_arch@)
+if(@depends_crosscompiling@)
```

Comment:
I really dislike that we still have to have this logic (and the depends counterpart), and I still don't really understand why it's needed. Now that the depends hacks are being removed, shouldn't we just be able to set `CMAKE_SYSTEM_NAME`, because the build system isn't going to operate specially for depends, and we've removed the `CMAKE_CROSSCOMPILING` usage?

In any case, according to the docs, if we are setting `CMAKE_SYSTEM_NAME`, we should also be setting `CMAKE_SYSTEM_VERSION`, or is that not needed?
> CMAKE_SYSTEM_NAME may be set explicitly when first configuring a new build tree in order to enable [cross compiling](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-toolchain). In this case the [CMAKE_SYSTEM_VERSION](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_VERSION.html#variable:CMAKE_SYSTEM_VERSION) variable must also be set explicitly.

Reply:
> we've removed the `CMAKE_CROSSCOMPILING` usage

Right. But CMake internally still uses it, which might affect its behaviour (feature tests, compiler invocation string, testing framework etc).

Reply:
> In any case, according to the docs, if we are setting `CMAKE_SYSTEM_NAME`, we should also be setting `CMAKE_SYSTEM_VERSION`

FWIW, we are not doing this now: https://github.com/hebasto/bitcoin/blob/7fcf4e99791ca5be0b068ac03a81a50ece11dba3/depends/funcs.mk#L193-L197

> or is that not needed?

From the CMake source code, it follows that it used for Android and Darwin platforms. For example, on Darwin, it affects on which flags will be used by default (`-Wl,-search_paths_first`).

Reply:
> For example, on Darwin, it affects on which flags will be used by default (-Wl,-search_paths_first).

This seems like a more correct thing to do, so that's fine, but apparently it will have no effect with lld.

Reply:
> I really dislike that we still have to have this logic (and the depends counterpart), and I still don't really understand why it's needed. Now that the depends hacks are being removed, shouldn't we just be able to set `CMAKE_SYSTEM_NAME`, because the build system isn't going to operate specially for depends, and we've removed the `CMAKE_CROSSCOMPILING` usage?

FWIW, in depends we do:
```bash
cross_compiling=maybe
host_alias="@HOST@"
```

Which I believe pretty much has the same effect for autotools that @hebasto is describing here for CMake.

Reply:
> Which I believe pretty much has the same effect for autotools that @hebasto is describing here for CMake.

I think the difference here is we've created something special called `depends_crosscompiling`, which is just done by checking if host != build in depends. Given how this is currently implemented, I assume it only fixes the issue (cross-compiling unnecessarily) for depends builds, and if anyone passed a non-depends toolchain file, which had, for example:
```bash
  set(CMAKE_SYSTEM_NAME name)
  set(CMAKE_SYSTEM_VERSION version)
```

Then they'd get whatever CMake behaviour we are trying to avoid here? Is that something we care about? If yes, this workaround should exist in the main build system, so it applies to all builds, and not be special to depends.

Reply:
>... if anyone passed a non-depends toolchain file, which had, for example:
> 
> ```shell
>   set(CMAKE_SYSTEM_NAME name)
>   set(CMAKE_SYSTEM_VERSION version)
> ```
> 
> Then they'd get whatever CMake behaviour we are trying to avoid here?

Yes, they would.

> Is that something we care about?

No, it is not. 

It is explicitly [documented](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html) that the `CMAKE_CROSSCOMPILING` variable:
> will be set to true by CMake if the [CMAKE_SYSTEM_NAME](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html#variable:CMAKE_SYSTEM_NAME) variable has been set manually.

It is one's responsibility who sets `CMAKE_SYSTEM_NAME`.

Reply:
Then we should at least be documenting why we are avoiding this behaviour (besides calling something that isn't cross-compiling, cross-compiling, which is just confusing).

Reply:
Sorry, but I don't understand this point. We are not trying to avoid something, rather following the CMake docs.

When we want to enable cross-compiling, we set the `CMAKE_SYSTEM_NAME` variable, which sets the `CMAKE_CROSSCOMPILING` one explicitly.

When we don't want to enable cross-compiling, we skip touching the `CMAKE_SYSTEM_NAME` variable.

Anyway, I'll be happy to add a comment to this code. Do you mind suggesting a good wording for it?

Reply:
> When we don't want to enable cross-compiling, we skip touching the CMAKE_SYSTEM_NAME variable.

This is what we are avoiding. Otherwise CMake decides things are cross-compilation, even when they are not. Otherwise we'd be free to set this all the time.

Reply:
@hebasto, for the wording, if @fanquake allows me here and agrees with it, perhaps something like:
```
# Setting CMAKE_SYSTEM_NAME only when cross-compiling is intended.
# This prevents CMake from misinterpreting the build as a cross-compilation,
# which could lead to incorrect configurations. If we are not cross-compiling,
# leave CMAKE_SYSTEM_NAME unset to allow CMake to correctly infer the build system.
```
If it's too long, I'd just get the first 2 lines and add the link to the `CMAKE_CROSSCOMPILING` [documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html).

Reply:
> Then we should at least be documenting why we are avoiding this behaviour (besides calling something that isn't cross-compiling, cross-compiling, which is just confusing).

An explanatory comment has been added.